ai_edge_torch/generative/layers/attention.py

```diff
@@ -159,28 +159,35 @@ def forward(
     # Assemble into a number of query groups to support MHA, MQA and GQA.
     q_per_kv = self.config.num_heads // self.config.num_query_groups
     total_qkv = q_per_kv + 2  # Each group has >=1 queries, 1 key, and 1 value.
-    qkv = qkv.view(
-        B, T, self.config.num_query_groups, total_qkv, self.head_dim
-    )  # (B, T, num_query_groups, total_qkv, head_dim)
 
-    # Split batched computation into three.
-    q, k, v = qkv.split((q_per_kv, 1, 1), dim=-2)
+    if self.config.qkv_transpose_before_split:
+      qkv = qkv.view(
+          B, T, total_qkv, self.config.num_query_groups, self.head_dim
+      )  # (B, T, total_qkv, num_query_groups, head_dim)
+      q, k, v = qkv.split((q_per_kv, 1, 1), dim=-3)
+    else:
+      qkv = qkv.view(
+          B, T, self.config.num_query_groups, total_qkv, self.head_dim
+      )  # (B, T, num_query_groups, total_qkv, head_dim)
+      # Split batched computation into three.
+      q, k, v = qkv.split((q_per_kv, 1, 1), dim=-2)
```

Comment:
it seems the only difference between the two cases is the order of `total_qkv` in the first view operation. After split, the q/k/v tensors should be same for both cases? If so, can we just keep one case here?

Reply:
Do you mean something like this? keep the split operation into a single.

```python
if self.config.qkv_transpose_before_split:
  qkv = qkv.view(
      B, T, total_qkv, self.config.num_query_groups, self.head_dim
  )  # (B, T, total_qkv, num_query_groups, head_dim)
  qkv_axis = -3
else:
  qkv = qkv.view(
      B, T, self.config.num_query_groups, total_qkv, self.head_dim
  )  # (B, T, num_query_groups, total_qkv, head_dim)
  qkv_axis = -2
  
# Split batched computation into three.
q, k, v = qkv.split((q_per_kv, 1, 1), dim=qkv_axis)
q = q.reshape(B, T, -1, self.head_dim)
k = k.reshape(B, T, -1, self.head_dim)
v = v.reshape(B, T, -1, self.head_dim)
```

Reply:
discussed offline, closing for now.